keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -320,28 +344,54 @@ class ApplicationService(
     environments: Set<ArtifactSummaryInEnvironment>
   ): ArtifactVersionSummary {
-    val artifactSupplier = artifactSuppliers.supporting(artifact.type)
-    val releaseStatus = repository.getReleaseStatus(artifact, version)
-    val artifactInstance = repository.getArtifactInstance(artifact.name, artifact.type, version, releaseStatus)
-      ?: throw InvalidSystemStateException("Loading artifact version $version failed for known artifact $artifact.")
 
+    val artifactSupplier = artifactSuppliers.supporting(artifact.type)
+    val artifactInstance = getArtifactInstance(artifact, version)
+            ?: throw InvalidSystemStateException("Loading artifact version $version failed for known artifact $artifact.")
     return ArtifactVersionSummary(
-      version = version,
-      environments = environments,
-      displayName = artifactSupplier.getVersionDisplayName(artifactInstance),
-      createdAt = artifactInstance.createdAt,
-
-      // first attempt to use the artifact metadata fetched from the DB, then fallback to the default if not found
-      build = artifactInstance.buildMetadata
-        ?: artifactSupplier.parseDefaultBuildMetadata(artifactInstance, artifact.versioningStrategy),
-      git = artifactInstance.gitMetadata
-        ?: artifactSupplier.parseDefaultGitMetadata(artifactInstance, artifact.versioningStrategy)
-    )
-  }
+        version = version,
+        environments = environments,
+        displayName = artifactSupplier.getVersionDisplayName(artifactInstance),
+        createdAt = artifactInstance.createdAt,
+
+        // first attempt to use the artifact metadata fetched from the DB, then fallback to the default if not found
+        build = artifactInstance.buildMetadata
+          ?: artifactSupplier.parseDefaultBuildMetadata(artifactInstance, artifact.versioningStrategy),
+        git = artifactInstance.gitMetadata
+          ?: artifactSupplier.parseDefaultGitMetadata(artifactInstance, artifact.versioningStrategy)
+      )
+    }
 
   fun getApplicationEventHistory(application: String, limit: Int) =
     repository.applicationEventHistory(application, limit)
 
   private fun ConstraintState.toConstraintSummary() =
     StatefulConstraintSummary(type, status, createdAt, judgedBy, judgedAt, comment, attributes)
+
+  private fun getArtifactInstance(
+    artifact: DeliveryArtifact,
+    version: String
+  ): PublishedArtifact? {
+    val releaseStatus = repository.getReleaseStatus(artifact, version)
+    return repository.getArtifactInstance(artifact.name, artifact.type, version, releaseStatus)
+  }
+
+  // Generating a stash diff link between source and target versions (the order does matter!)
+  private fun generateDiffLink (newerGitMetadata: GitMetadata?, olderGitMetadata: GitMetadata?) : String? {
+    return if (newerGitMetadata !=null && olderGitMetadata != null) {
+      "https://stash.corp.netflix.com/projects/${newerGitMetadata.project}/repos/${newerGitMetadata.repo?.name}/compare/diff?" +
```

Comment:
This base URL cannot be hard-coded here, but instead retrieved at runtime to reflect what is configured in the Spinnaker installation for this particular artifact. 

Igor has an endpoint you can call to retrieve that information:
```
GET https://{{igor.host}}/scm/masters
{
  "stash": "<url>",
  "gitHub": "<url>"
}
```
The code might look something like this:
```kotlin
// The `scmType` field is currently missing from the model and is something you'll have to add/fill in
val baseUrl =  igorService.getScmUrl(newerGitMetadata.scmType)
when (newerGitMetadata.scmType) {
  "stash" -> "$baseUrl/projects/${newerGitMetadata.project}/repos/${newerGitMetadata.repo?.name}/compare/diff?" // ...
  "github" -> // ...
  else -> throw UnsupportedScmType(newerGitMetadata.scmType)
}
```
P.S. Whenever you find yourself using an internal Netflix URL like you have here, that's generally a good sign to take a pause and reevaluate, including to avoid leaking potentially sensitive information. 🙂

Reply:
Thanks for the detailed comment! I didn't think using a stash link is considered as sensitive information :) but I totally agree it can be better! will make the change.

Reply:
@lpollo you’ve suggested that I’ll use igor to fetch the base url, but I’m having a dependencies issue as the applicationService code is in keel-core and igor is in keel-igor , so I can’t add igor to that module or I’ll have a circled dependency.. no?

Reply:
fixed!